### PR TITLE
Add CTA when widget is empty

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/392891325557410/1195019227318197
Tech Design URL: 
CC:

**Description**:

* Adds CTA to the favorites widget when the user has no favorites.  
* Also tries to reuse existing tabs when possible when launching.
* Allows existing bookmarks to be added as favorites and visa versa

***Approvals***:

* Copy approved.
* Design review approved.
* Product review in progress.

**Steps to test this PR**:

***Scenario: Standard flow***:
1. Install and run app and get through onboarding
2. Add favorites widget
3. Tap add favorite > from widget
4. Dax should tell you it's easy to create a short cut
5. Load a page (either a search or a page), menu should have pulsing animation
6. Tap menu icon, menu should have "add favorite" with hand waving emoji next to it

***Scenario: Did not complete onboarding***:
1. Install app, then kill it without starting the onboarding.  
1. Add the favorites widget
1. Tap add favorite > from widget
1. App should launch, run through onboarding then drop in to the add favorite flow
1. Once completed the dax dialog onboarding should continue from the next time you browse to a page

***Scenario: Favorite already open in tab***:
1. Launch app and load a page
1. Tap add favorite > from widget
1. App should launch and start add favorite flow
1. Instead of browsing to a page, use the tab switcher to open existing page
1. Menu should have pulsing animation
1. Complete add favorite flow as normal

***Scenario: App killed during flow***:
1. Tap add favorite > from widget
1. App should launch and start add favorite flow
1. Kill the app
1. Launch the app, flow should not continue

***Scenario: Fire button during flow***:
1. Tap add favorite > from widget
1. App should launch and start add favorite flow
1. Use the Fire button
1. Add favorite flow should not continue

***Reuse home tab***:
1. Tapping the search CTA on the widgets will select first existing home tab or create a new one.

***Reuse open tab***:
1. Launching a favorite or handling a default browser will re-use an existing tab if its URL matches exactly.

***Bookmarks/favorites***:
1. Browse to a new page
1. Open the menu - both add bookmark and add favorite are present
1. Add bookmark
1. Open the menu - add favorite is present
1. Add favorite
1. Open the menu - neither option is present
1. Repeat starting with adding a favorite and checking the bookmark menu item appears

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [x] Portrait
* [x] Landscape

**Device Testing**:

* [x] iPhone SE (1st Gen)
* [x] iPhone 8
* [x] iPhone X
* [x] iPad

**OS Testing**:

* [x] iOS 11
* [x] iOS 12
* [x] iOS 13

**Theme Testing**:

* [x] Light theme
* [x] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Comment by brindy** [comment]
Hey @bwaresiak @samsymons @THISISDINOSAUR - If the design and product review go well I'm looking to release this on Monday so could do with someone giving it a look today if possible please! 🙏

**Comment by brindy** [comment]
That should be the last of the changes from the design review - have updated the description to include the new change.

**Comment by brindy** [comment]
Had some feedback from the Product Review regarding copy (needs another pass) and some launch glitch and @bwaresiak has found a few things as well so please consider this WIP for now and I'll look to release this after the other stuff has gone out.

**Review by bwaresiak — APPROVED** [review]
LGTM!

File: DuckDuckGo/TabManager.swift [hunk 1]
```diff
@@ -123,7 +123,21 @@ class TabManager {
...
         save()
     }
-    
+
+    func firstHomeTab() -> Tab? {
+        return model.tabs.first(where: { $0.link == nil })
+    }
+
+    func first(withUrl url: URL) -> Tab? {
+        return model.tabs.first(where: { $0.link?.url == url })
```
Comment (samsymons) [inline DuckDuckGo/TabManager.swift]: An edge case here is if you've opened a site that is then upgraded to HTTPS, and later open the HTTP link again. Is it worth reusing the HTTPS tab?

Comment (brindy) [reply DuckDuckGo/TabManager.swift]: Good point - I'll fix that before merging.

Comment (brindy) [reply DuckDuckGo/TabManager.swift]: I've just pushed that.  Any chance you could give that a wee test, please? 

Also, I made a small fix to improve favicon loading for favorites, so if you could keep an eye on that while you do that, that would be great, thanks!

Comment (samsymons) [reply DuckDuckGo/TabManager.swift]: This is looking good to me now, nice work @brindy!

I found one new edge case where I tested with a local restaurant's website which I opened via Apple Maps, but the HTTPS redirected version adds a trailing forward slash, so it still fails.

```
(lldb) po linkUrl
▿ https://www.thevillageidiot.ca/
  - _url : https://www.thevillageidiot.ca/

(lldb) po url
▿ http://www.thevillageidiot.ca
  - _url : http://www.thevillageidiot.ca
```

I think this is getting into edge case of edge case territory, so I'm honestly fine with the current behaviour.

Comment (brindy) [reply DuckDuckGo/TabManager.swift]: Hah, OK thanks.  It's better than it was for sure!

**Review by samsymons — APPROVED** [review]
This is a great improvement, awesome work. 👏 

Interestingly widget previews seem to be unstable on this branch and won't update live for me, but I don't see anything in this PR that would've caused that. I'm betting it's just a local issue.

**Edit:** I also tested the debug menu on iPhone and iPad, working as expected!

:shipit:

**Comment by brindy** [comment]
Thanks @bwaresiak and @samsymons!

**Pull request merged**